File: practice/nodes/pub_gps.cpp
#include <ros/ros.h>
#include <sensor_msgs/NavSatFix.h>
#include <sensor_msgs/NavSatStatus.h>
#include <serial/serial.h>
#include <string>
#include <sstream>

serial::Serial ser;

void parseGPSData(const std::string &nmea_sentence, sensor_msgs::NavSatFix &fix_msg) {
    if (nmea_sentence.substr(0, 6) == "$GPGGA") {
        // NMEA 문장 파싱 로직 (생략)
    }
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "gps_publisher");
    ros::NodeHandle nh;

    ros::Publisher gps_pub = nh.advertise<sensor_msgs::NavSatFix>("/gps/fix", 1000);

    // 시리얼 포트 설정
    try {
        ser.setPort("/dev/ttyUSB0");
        ser.setBaudrate(115200);
        serial::Timeout to = serial::Timeout::simpleTimeout(1000);
        ser.setTimeout(to);
        ser.open();
    } catch (serial::IOException& e) {
        ROS_ERROR_STREAM("Unable to open port ");
        return -1;
    }

    if (ser.isOpen()) {
        ROS_INFO_STREAM("Serial Port initialized");
    } else {
        return -1;
    }

    ros::Rate loop_rate(5);  // 5Hz 주기로 GPS 데이터 수신

    while (ros::ok()) {
        if (ser.available()) {
            std::string gps_data = ser.readline();
            ROS_INFO_STREAM("Received GPS Data: " << gps_data);  // 수신된 GPS 데이터 로그 출력
            
            sensor_msgs::NavSatFix fix_msg;
            parseGPSData(gps_data, fix_msg);  // NMEA 데이터 파싱

            gps_pub.publish(fix_msg);
        }
        ros::spinOnce();
        loop_rate.sleep();
    }
}

File: practice/nodes/pub_gps.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import NavSatFix
import serial

def gps_publisher():
    rospy.init_node('gps_publisher', anonymous=True)
    pub = rospy.Publisher('gps/fix', NavSatFix, queue_size=10)

    # 아두이노와의 직렬 포트 설정
    ser = serial.Serial('/dev/ttyUSB0', 115200)
    rate = rospy.Rate(1)  # 1Hz

    while not rospy.is_shutdown():
        if ser.in_waiting > 0:
            data = ser.readline().decode('utf-8').strip()
            try:
                # 데이터 파싱
                lat_str, lon_str, alt_str = data.split()
                latitude = float(lat_str.split(':')[1])
                longitude = float(lon_str.split(':')[1])
                altitude = float(alt_str.split(':')[1])
                
                # NavSatFix 메시지 생성
                gps_msg = NavSatFix()
                gps_msg.latitude = latitude
                gps_msg.longitude = longitude
                gps_msg.altitude = altitude
                gps_msg.header.stamp = rospy.Time.now()
                pub.publish(gps_msg)
                rospy.loginfo("Published GPS data: %s", data)
            except ValueError:
                rospy.logwarn("Failed to parse GPS data: %s", data)

        rate.sleep()

if __name__ == '__main__':
    try:
        gps_publisher()
    except rospy.ROSInterruptException:
        pass

